qiskit_aer/backends/aer_compiler.py

```diff
@@ -621,16 +685,114 @@ def assemble_circuit(circuit: QuantumCircuit):
+            inst,
+            qubit_indices,
+            clbit_indices,
+            is_conditional,
+            conditional_reg,
+            conditional_expr,
         )
         index_map.append(num_of_aer_ops - 1)
 
     return aer_circ, index_map
 
 
-def _assemble_op(aer_circ, inst, qubit_indices, clbit_indices, is_conditional, conditional_reg):
+def _assemble_type(expr_type):
+    if isinstance(expr_type, Uint):
+        return AerUint(expr_type.width)
+    elif isinstance(expr_type, Bool):
+        return AerBool()
+    else:
+        raise AerError(f"unknown type: {expr_type.__class__}")
+
+
+def _assemble_clbit_indices(circ, c):
+    if isinstance(c, (ClassicalRegister, list)):
+        return [circ.find_bit(cbit).index for cbit in c]
+    elif isinstance(c, Clbit):
+        return [circ.find_bit(c).index]
+    else:
+        raise AerError(f"unknown clibt list: {c.__class__}")
+
+
+def _assemble_unary_operator(op):
+    if op is Unary.Op.BIT_NOT:
+        return AerUnaryOp.BitNot
+    elif op is Unary.Op.LOGIC_NOT:
+        return AerUnaryOp.LogicNot
+    else:
+        raise AerError(f"unknown op: {op}")
+
+
+_BINARY_OP_MAPPING = {
+    Binary.Op.BIT_AND: AerBinaryOp.BitAnd,
+    Binary.Op.BIT_OR: AerBinaryOp.BitOr,
+    Binary.Op.BIT_XOR: AerBinaryOp.BitXor,
+    Binary.Op.LOGIC_AND: AerBinaryOp.LogicAnd,
+    Binary.Op.LOGIC_OR: AerBinaryOp.LogicOr,
+    Binary.Op.EQUAL: AerBinaryOp.Equal,
+    Binary.Op.NOT_EQUAL: AerBinaryOp.NotEqual,
+    Binary.Op.LESS: AerBinaryOp.Less,
+    Binary.Op.LESS_EQUAL: AerBinaryOp.LessEqual,
+    Binary.Op.GREATER: AerBinaryOp.Greater,
+    Binary.Op.GREATER_EQUAL: AerBinaryOp.GreaterEqual,
+}
+
+
+def _assemble_binary_operator(op):
+    if op in _BINARY_OP_MAPPING:
+        return _BINARY_OP_MAPPING[op]
+    else:
+        raise AerError(f"unknown op: {op}")
+
+
+def _assemble_expr(circ, expr):
+    aer_expr = None
+    if expr is None:
+        aer_expr = None
+    elif isinstance(expr, Value):
+        if isinstance(expr.type, Uint):
+            aer_expr = AerUintValue(expr.type.width, expr.value)
+        elif isinstance(expr.type, Bool):
+            aer_expr = AerBoolValue(expr.value)
+        else:
+            raise AerError(f"invalid value type is specified: {expr.type.__class__}")
+    elif isinstance(expr, Var):
+        aer_expr = AerVar(_assemble_type(expr.type), _assemble_clbit_indices(circ, expr.var))
+    elif isinstance(expr, Cast):
+        aer_expr = AerCast(_assemble_type(expr.type), _assemble_expr(circ, expr.operand))
+    elif isinstance(expr, Unary):
+        aer_expr = AerUnaryExpr(
+            _assemble_unary_operator(expr.op), _assemble_expr(circ, expr.operand)
+        )
+    elif isinstance(expr, Binary):
+        aer_expr = AerBinaryExpr(
+            _assemble_binary_operator(expr.op),
+            _assemble_expr(circ, expr.left),
+            _assemble_expr(circ, expr.right),
+        )
+    else:
+        raise AerError(f"unknown expression: {expr.__class__}")
```

Comment:
Yes. `ExprVisitor` looks much better than the current my impl.

Reply:
@jakelishman I updated this PR to use `ExprVisitor`. I think this class does not have `visit()` so I created it. Maybe we can use `accept()` instead of creating this new method.

Reply:
Yeah `ExprVisitor` doesn't have a `visit` method because you call it as `expr.accept(visitor)`; the `Expr` subclasses handle choosing which method to call, rather than the `ExprVisitor` doing magic inference.